### PR TITLE
[sgl atom] enable qwen3.8 in sgl atom

### PR DESCRIPTION
**Pull request by ZLkanyo009**
# Enable Qwen3.8 and tc_piecewise prefill in sgl atom

## Motivation

Enable Qwen3.8-27B and Qwen3.8-2.4T on SGLang-ATOM while aligning its attention, KV-cache, CUDA Graph, and Inductor behavior with Native ATOM.

This also fixes correctness issues caused by padded CUDA Graph batches, stale attention metadata, and unstable tensor addresses between tc_piecewise graph segments.

### 1. Qwen3.8 model registration and runtime integration

- Register `Qwen3_5ForCausalLM` and `Qwen3_5MoeForCausalLM`.
- Extend Qwen3.5 prepare adaptations to CausalLM and MoE CausalLM architectures.
- Add Qwen3.5 construction, cache-view binding, and forward-metadata hooks.
- Preserve the existing DFLASH and explicit GDN context behavior.
- Propagate `save_kv_cache` through the SGLang plugin runtime.

### 2. Native ATOM attention bridge

Add `qwen3_5_attention.py` to translate SGLang runtime state into Native ATOM MHA/GDN inputs:

Previously, Qwen3.8 Full Attention in SGLang-ATOM still executed SGLang's
attention implementation through the Full Attention backend. This PR replaces
that compute path with Native ATOM attention while keeping SGLang responsible
for scheduling, request metadata, and KV-cache allocation.

- Reuse SGLang-owned KV and Mamba pools without copying storage.
- Build ATOM-compatible prefill and decode metadata.
- Support cached-prefix and cold-prefill paths.
- Bind K/V cache using the actual SGLang page geometry.
- Cache KV/Mamba views and invalidate them when pool identities change.
- Combine dense MHA and GDN metadata in the same ATOM forward context.

### 3. MHA and FP8 prefill optimization

This part changes the shared `atom/model_ops/attention_mha.py` implementation,
so it benefits Native ATOM as well as SGLang-ATOM. For eligible Native ATOM
Qwen3.8 pure-prefill workloads, it replaces the BF16 varlen attention path with
AITER's per-tensor FP8 FMHA path, reducing attention compute and memory traffic
and improving prefill performance.

- Add the AITER per-tensor FP8 prefill attention path for gfx950, head dimension 256, and FP8 KV cache.
- Preallocate the FP8 attention scale outside CUDA Graph capture.
- Preserve `logits_soft_cap` for dispatch eligibility checks.
- Select the ASM cache layout according to the actual external KV-cache shape.
- Add `ATOM_AITER_FP8_PREFILL_ATTN` as a rollback switch.

The optimization is enabled only when the workload satisfies the validated
constraints: gfx950, head dimension 256, FP8 KV cache, pure prefill without a
cached prefix, no sliding window or attention sinks, and no logits soft cap or
dropout. Other workloads retain the existing fallback path.

#### 4.1 CUDA Graph correctness

Both prefill and decode CUDA Graph paths had the same class of correctness
problem: synthetic padding rows/tokens could be interpreted as valid work and
read or write real KV/GDN state slots. This PR fixes the no-op sentinel contract
for both phases.

- Decode: mark padded `kv_lens` rows as empty so graph bucket padding cannot
  overwrite request-pool row 0.
- Prefill: trim attention/GDN execution to the real token count and zero the
  padded output range.
- Prefill and decode: set padded GDN state indices to `-1`.
- Prefill and decode: build zero-length query ranges for padded requests.
- Translate `save_kv_cache=False` to the established `slot_mapping=-1` no-write contract.

The prefill tc_piecewise path additionally keeps attention/GDN outside captured
dense segments and writes into graph-stable output buffers, so replay uses
current metadata without changing the next segment's captured input address.

#### 4.2 tc_piecewise split boundaries

Register both Native returning ops and Qwen3.5 stable-output ops as SGLang split boundaries:

- `aiter.unified_attention_with_output_base`
- `aiter.linear_attention_with_output_base`
- `aiter.sglang_qwen35_attention_with_stable_output`
- `aiter.sglang_qwen35_linear_attention_with_stable_output`

Attention and GDN remain outside compiled dense segments so every replay reads the current batch metadata.

For real prefill tc_piecewise CUDA Graphs, the Qwen3.5-specific ops write into preallocated output buffers. This keeps the following graph segment's input address identical between capture and replay.

Compile-only prefill, eager execution, and full decode CUDA Graphs continue to use the Native returning ops.

#### 4.3 Optional prefill compile-only mode

Add:

```bash
ATOM_SGLANG_PREFILL_COMPILE_ONLY=1
```

This mode:

- Keeps tc_piecewise FX splitting and dynamic-shape Inductor compilation.
- Disables per-bucket prefill CUDA Graph capture.
- Avoids bucket padding and static input staging.
- Executes directly against the live `ForwardBatch`.
- Enables `enable_auto_functionalized_v2` only in the compile-only prefill scope, restoring the GDN reduction/pointwise fusion used by Native ATOM.
- Temporarily selects the Inductor compiler without permanently mutating SGLang server configuration.

Without this environment variable, the normal tc_piecewise CUDA Graph path remains active.

#### 4.4 Triton kernel lifetime

Add a Triton kernel-retention patch for ROCm graph capture:

- Retain `CompiledKernel` objects whose destructor runs during capture or compile-only execution.
- Prevent `hipModuleUnload` from invalidating function handles referenced by captured graphs.
- Wrap SGLang's original `freeze_gc` implementation instead of replacing its GC policy.
- Patch runner-local `freeze_gc` aliases that were imported by value.

### 5. EAGLE3 runtime guard

Prevent ordinary generation payloads from entering the EAGLE3-only FutureMap state path.

This removes an unnecessary CPU-to-GPU state copy that synchronizes the forward stream after prefill while preserving the existing EAGLE3 payload behavior.

### 6. Code Re-organization

Group SGLang runtime patches under:

```text
atom/plugin/sglang/patches/
├── graph_capture_patch.py
├── prefill_compile_only_patch.py
└── triton_kernel_retention_patch.py
```

Move the EAGLE3 lifecycle test into `tests/plugin/`, matching the production module it validates.

## New environment variables

```text
ATOM_AITER_FP8_PREFILL_ATTN
ATOM_SGLANG_PREFILL_COMPILE_ONLY
ATOM_SGLANG_EAGLE3_TP_VERIFY_BROADCAST
```

## Command
ATOM:
```
#!/usr/bin/env bash
set -euo pipefail

MODEL_PATH=${MODEL_PATH:-/models/Qwen3.8-2.4T-A95B-Quark-MXFP4}

HOST=${HOST:-0.0.0.0}
PORT=${PORT:-8001}
GPU_IDS=${GPU_IDS:-0,1,2,3,4,5,6,7}
TP_SIZE=${TP_SIZE:-8}
MAX_NUM_SEQS=${MAX_NUM_SEQS:-256}
GPU_MEMORY_UTILIZATION=${GPU_MEMORY_UTILIZATION:-0.90}

export CUDA_VISIBLE_DEVICES="${GPU_IDS}"
export HIP_VISIBLE_DEVICES="${GPU_IDS}"
export PYTHONPATH="${ATOM_PATH}${PYTHONPATH:+:${PYTHONPATH}}"

export AITER_LOG_LEVEL=${AITER_LOG_LEVEL:-INFO}
export TORCHINDUCTOR_COMPILE_THREADS=${TORCHINDUCTOR_COMPILE_THREADS:-128}
export ATOM_AITER_FP8_PREFILL_ATTN=${ATOM_AITER_FP8_PREFILL_ATTN:-1}


exec python3 -m atom.entrypoints.openai_server \
    --model "${MODEL_PATH}" \
    --host "${HOST}" \
    --server-port "${PORT}" \
    --trust-remote-code \
    -tp "${TP_SIZE}" \
    --max-num-seqs "${MAX_NUM_SEQS}" \
    --gpu-memory-utilization "${GPU_MEMORY_UTILIZATION}" \
    --kv_cache_dtype fp8 \
    --no-enable_prefix_caching \
    --cudagraph-mode FULL
```

SGL ATOM:
```
#!/usr/bin/env bash
set -euo pipefail

MODEL_PATH=${MODEL_PATH:-/models/Qwen3.8-2.4T-A95B-Quark-MXFP4}

HOST=${HOST:-0.0.0.0}
PORT=${PORT:-8001}
GPU_IDS=${GPU_IDS:-0,1,2,3,4,5,6,7}
TP_SIZE=${TP_SIZE:-8}
MAX_RUNNING_REQUESTS=${MAX_RUNNING_REQUESTS:-256}
MEM_FRACTION_STATIC=${MEM_FRACTION_STATIC:-0.90}
LOG_FILE=${LOG_FILE:-qwen38-2.4t-sglang-atom-tp8.log}

export SGLANG_PLUGINS=atom_sglang
export SGLANG_EXTERNAL_MODEL_PACKAGE=atom.plugin.sglang.models
export SGLANG_EXTERNAL_MM_PROCESSOR_PACKAGE=atom.plugin.sglang.models
export SGLANG_USE_AITER=1
export AITER_LOG_LEVEL=${AITER_LOG_LEVEL:-INFO}
export ATOM_AITER_FP8_PREFILL_ATTN=${ATOM_AITER_FP8_PREFILL_ATTN:-1}

export ATOM_SGLANG_PREFILL_COMPILE_ONLY=${ATOM_SGLANG_PREFILL_COMPILE_ONLY:-1}
export TORCHINDUCTOR_COMPILE_THREADS=${TORCHINDUCTOR_COMPILE_THREADS:-128}

TORCH_COMPILE_MAX_BS=${TORCH_COMPILE_MAX_BS:-64}
PREFILL_CAPTURE_MAX_TOKENS=${PREFILL_CAPTURE_MAX_TOKENS:-16384}
PREFILL_CAPTURE_TOKEN_BUCKETS=${PREFILL_CAPTURE_TOKEN_BUCKETS:-"8192 16384"}
read -r -a PREFILL_CAPTURE_TOKEN_LIST <<< "${PREFILL_CAPTURE_TOKEN_BUCKETS}"

export SGLANG_MAMBA_SSM_DTYPE=bfloat16

exec python3 -m sglang.launch_server \
    --model-path "${MODEL_PATH}" \
    --host "${HOST}" \
    --port "${PORT}" \
    --trust-remote-code \
    --tensor-parallel-size "${TP_SIZE}" \
    --attention-backend aiter \
    --kv-cache-dtype fp8_e4m3 \
    --mem-fraction-static "${MEM_FRACTION_STATIC}" \
    --max-running-requests "${MAX_RUNNING_REQUESTS}" \
    --disable-radix-cache \
    --page-size 16 \
    --enable-torch-compile \
    --torch-compile-max-bs "${TORCH_COMPILE_MAX_BS}" \
    --cuda-graph-backend-prefill tc_piecewise \
    --cuda-graph-max-bs-prefill "${PREFILL_CAPTURE_MAX_TOKENS}" \
    --cuda-graph-bs-prefill "${PREFILL_CAPTURE_TOKEN_LIST[@]}" \
    --cuda-graph-tc-compiler inductor

```

## Acc
ATOM && SGL ATOM：
```
lm_eval --model local-completions \
--model_args '{"base_url":"http://localhost:8001/v1/completions","model":"/models/Qwen3.8-2.4T-A95B-Quark-MXFP4","num_concurrent":256,"max_retries":10,"max_gen_toks":2048,"max_length":8192,"tokenized_requests":false,"timeout":3600}' \
--tasks gsm8k \
--num_fewshot 5 \
--trust_remote_code
```

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.953|±  |0.0058|
|     |       |strict-match    |     5|exact_match|↑  |0.953|±  |0.0058|

```

```
#!/usr/bin/env bash
set -euo pipefail

MODEL_PATH=${MODEL_PATH:-/models/Qwen3.8-2.4T-A95B-Quark-MXFP4}
BASE_URL=${BASE_URL:-http://localhost:8001/v1/chat/completions}
NUM_CONCURRENT=${NUM_CONCURRENT:-64}
OUTPUT_PATH=${OUTPUT_PATH:-./qwen38-gsm8k-results}

lm_eval --model local-chat-completions \
  --model_args "model=${MODEL_PATH},base_url=${BASE_URL},num_concurrent=${NUM_CONCURRENT},max_retries=3,timeout=1200,max_gen_toks=8192,tokenized_requests=False,trust_remote_code=True" \
  --tasks gsm8k \
  --num_fewshot 5 \
  --batch_size 1 \
  --output_path "${OUTPUT_PATH}" \
  --gen_kwargs 'do_sample=True,temperature=1.0,top_p=0.95,top_k=20,min_tokens=1,reasoning_effort=low' \
  --apply_chat_template \
  --fewshot_as_multiturn \
  --trust_remote_code
```

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.981|±  |0.0038|
|     |       |strict-match    |     5|exact_match|↑  |0.981|±  |0.0038|
```
## Performance
```
#!bin/bash

MODEL=/models/Qwen3.8-2.4T-A95B-Quark-MXFP4
RANGE_RATIO=1.0
 

ISL=8192
OSL=64
CON=64
NUM=$(( CON * 3 ))

echo "ATOM Model=${MODEL}"
echo "ATOM ISL=${ISL}, OSL=${OSL}, NUM=${NUM}, CON=${CON} RANGE_RATIO=${RANGE_RATIO}"
 
sleep 2
 
# git clone https://github.com/kimbochen/bench_serving.git
echo "Starting bench with model: ${MODEL}"
python bench_serving/benchmark_serving.py \
    --model=$MODEL \
    --backend=sglang \
    --base-url=http://localhost:8001 \
    --dataset-name=random \
    --random-input-len=$ISL \
    --random-output-len=$OSL \
    --random-range-ratio ${RANGE_RATIO} \
    --num-prompts=${NUM} \
    --max-concurrency=${CON} \
    --request-rate=inf \
    --ignore-eos \
    --save-result \
    --percentile-metrics="ttft,tpot,itl,e2el" \
    --result-dir=./

echo "ATOM Model=${MODEL}"
echo "ATOM ISL=${ISL}, OSL=${OSL}, NUM=${NUM}, CON=${CON}"
rm -rf ./*.json
```

### ATOM
before mha kernel change:
```
============ Serving Benchmark Result ============
Successful requests:                     192       
Benchmark duration (s):                  74.64     
Total input tokens:                      1572864   
Total generated tokens:                  12288     
Request throughput (req/s):              2.57      
Output token throughput (tok/s):         164.64    
Total Token throughput (tok/s):          21238.70  
---------------Time to First Token----------------
Mean TTFT (ms):                          12224.33  
Median TTFT (ms):                        12147.62  
P99 TTFT (ms):                           23446.89  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          200.57    
Median TPOT (ms):                        200.40    
P99 TPOT (ms):                           379.73    
---------------Inter-token Latency----------------
Mean ITL (ms):                           197.44    
Median ITL (ms):                         27.41     
P99 ITL (ms):                            7460.32   
----------------End-to-end Latency----------------
Mean E2EL (ms):                          24860.28  
Median E2EL (ms):                        24565.98  
P99 E2EL (ms):                           25476.97  
==================================================
```

after:
```
============ Serving Benchmark Result ============
Successful requests:                     192       
Benchmark duration (s):                  72.20     
Total input tokens:                      1572864   
Total generated tokens:                  12288     
Request throughput (req/s):              2.68      
Output token throughput (tok/s):         170.19    
Total Token throughput (tok/s):          21954.98  
---------------Time to First Token----------------
Mean TTFT (ms):                          11601.42  
Median TTFT (ms):                        11609.74  
P99 TTFT (ms):                           22140.70  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          197.60    
Median TPOT (ms):                        197.42    
P99 TPOT (ms):                           371.88    
---------------Inter-token Latency----------------
Mean ITL (ms):                           194.52    
Median ITL (ms):                         29.36     
P99 ITL (ms):                            7241.70   
----------------End-to-end Latency----------------
Mean E2EL (ms):                          24050.42  
Median E2EL (ms):                        24053.77  
P99 E2EL (ms):                           24074.28  
==================================================
```

### SGL ATOM
before optimization‌：
```
============ Serving Benchmark Result ============
Successful requests:                     192       
Benchmark duration (s):                  83.69     
Total input tokens:                      1572864   
Total generated tokens:                  12288     
Request throughput (req/s):              2.29      
Output token throughput (tok/s):         146.83    
Total Token throughput (tok/s):          18941.32  
---------------Time to First Token----------------
Mean TTFT (ms):                          14575.07  
Median TTFT (ms):                        14591.64  
P99 TTFT (ms):                           29980.08  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          211.16    
Median TPOT (ms):                        211.11    
P99 TPOT (ms):                           396.33    
---------------Inter-token Latency----------------
Mean ITL (ms):                           211.16    
Median ITL (ms):                         31.91     
P99 ITL (ms):                            8356.76   
----------------End-to-end Latency----------------
Mean E2EL (ms):                          27878.05  
Median E2EL (ms):                        25696.48  
P99 E2EL (ms):                           32287.61  
==================================================
```

after:
```
============ Serving Benchmark Result ============
Successful requests:                     192       
Benchmark duration (s):                  71.39     
Total input tokens:                      1572864   
Total generated tokens:                  12288     
Request throughput (req/s):              2.69      
Output token throughput (tok/s):         172.12    
Total Token throughput (tok/s):          22203.82  
---------------Time to First Token----------------
Mean TTFT (ms):                          11542.29  
Median TTFT (ms):                        11536.80  
P99 TTFT (ms):                           22001.18  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          194.23    
Median TPOT (ms):                        193.81    
P99 TPOT (ms):                           361.04    
---------------Inter-token Latency----------------
Mean ITL (ms):                           194.23    
Median ITL (ms):                         27.18     
P99 ITL (ms):                            7780.75   
----------------End-to-end Latency----------------
Mean E2EL (ms):                          23779.04  
Median E2EL (ms):                        23773.99  
P99 E2EL (ms):                           23848.06  
==================================================
```

### CODE REVIEW

**Review by valarLip — DISMISSED** [review]

**Comment by valarLip** [comment]
conflict..

**Review by valarLip — APPROVED** [review]
